### ISSUE TITLE
Update kit-deployer to change labels after the deployment is complete

### ISSUE THREAD

**Issue by brianhall** [InVisionApp/kit-deployer#81]
From: [Quick Rollbacks](https://invisionapp.quip.com/npgPANELpg94)

tl;dr We want the ability to continue running the previous X number of deployments even after a new deployment is out. We will direct all traffic to the new deployment via labels allowing us to do an instant rollback by updating labels to send traffic back to a previous deployment.

# Deployment Workflow

1. If `strategy == fast-rollback` then continue
1. Rename deployment to `<name>-<sha>` to prevent updating the current live deployment
1. Deploy as usual
1. If deploy fails then skip (fail as usual), otherwise continue
1. Run `kubectl label --overwrite pod -l sha!=$sha3 state=backup` to archive the previous pods
1. Run `kubectl label --overwrite deployment -l sha!=$sha3 state=backup` to archive the previous deployment
1. Delete n-3 deployments that are older than sha=$sha3 deployment creationTimestamp

# Rollback Workflow

1. If `strategy == fast-rollback` then continue
1. Run `kubectl get deployment -l sha=$sha2` to get the deployment we are rolling back to
1. If no deployment is found, skip and do a deploy as usual, otherwise continue
1. Run `kubectl label --overwrite pod -l sha=$sha2 state=live` to unarchive the previous pods
1. Run `kubectl label --overwrite deployment -l sha=$sha2 state=live` to unarchive the previous deployment
1. Delete all deployments newer than sha=$sha2 deployment creationTimestamp

# TODO

- [x] Implement strategy lifecycle framework
- [x] Create `rolling-update` strategy as default strategy (will do nothing special)
- [x] Create `fast-rollback` strategy that will do the new deployment/rollback workflow

**Comment by chesleybrown** [comment]
I've updated the description with more details.

**Comment by chesleybrown** [comment]
We've since had to change the workflow. Here is the updated workflow:

# Deployment/Rollback Workflow

1. If `strategy == fast-rollback` then continue
1. Rename deployment to `<name>-<deployId>` to prevent updating the current live deployment
1. Suppress deploying any `service` manifests until after all deployment manifests are available
1. If `<name>-<deployId>` is not already deployed, then Deploy as usual, otherwise skip deployment
1. If deploy fails then skip (fail as usual), otherwise continue
1. Verify deployment is fully available
1. Verify `service` selectors does select pods running in the cluster
1. Deploy all service manifests
1. If NOT rollback, delete n-3 deployments that are older than `deployId=$deployId` deployment `creationTimestamp`
1. If rollback, delete all deployments newer than deploy=$deployId deployment `creationTimestamp`

**Comment by brianhall** [comment]
@chesleybrown Before you close this PR, I'd like that text (or whatever the final result is) in the Quip doc, por favor. That's super useful. ^

**Comment by brianhall** [comment]
One other thing, @chesleybrown - we'd discussed storing 4 backup versions to make the deployment of previous versions easier. From my memory, we were going to delete N-4, so that when the following occurs:

A - backup
B - backup
C - backup
D - backup
E - current
F - new

A would be deleted right away during deployment, and then if F is rolled back, there are still 3 version to roll back to in backup.

~A - backup~
B - backup
C - backup
D - backup
E - current
~F - new~ (rolled back)

Does that match your understanding?

**Comment by chesleybrown** [comment]
@brianhall I've updated the document with the latest workflow/diagrams.

**Comment by chesleybrown** [comment]
This was implemented in #85 and #91, going to close.
